### PR TITLE
Injection into test method parameters in Arquillian tests

### PR DESCRIPTION
**Pull request by jmartisk**
#3943
@mkouba @kenfinnigan @manovotn please have a look. I don't feel very comfortable with the `convertToTCCL` thingy but for some reason it was necessary (Arquillian is passing the method metadata to `TestEnricher` and `TestMethodExecutor` loaded by system class loader). Any better solution to go with?

### CODE REVIEW

File: test-framework/arquillian/src/main/java/io/quarkus/arquillian/InjectionEnricher.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        return values;
+    }
+
+    @SuppressWarnings("unchecked")
+    private <T> T getInstanceByType(BeanManager manager, final int position, final Method method) {
+        CreationalContext<?> cc = getCreationalContext();
+        return (T) manager.getInjectableReference(new MethodParameterInjectionPoint<T>(method, position, manager), cc);
+    }
+
+    private boolean isInjectClassPresent() {
```
Comment (mkouba) [inline test-framework/arquillian/src/main/java/io/quarkus/arquillian/InjectionEnricher.java]: There is no need to do this check.

File: test-framework/arquillian/src/main/java/io/quarkus/arquillian/InjectionEnricher.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+ */
+public class InjectionEnricher implements TestEnricher {
+    private static final String ANNOTATION_NAME = "javax.inject.Inject";
+
+    private static final Logger log = Logger.getLogger(TestEnricher.class.getName());
+
+    @SuppressWarnings("rawtypes")
+    @Inject
+    @TestScoped
+    private InstanceProducer<CreationalContext> creationalContextProducer;
```
Comment (mkouba) [inline test-framework/arquillian/src/main/java/io/quarkus/arquillian/InjectionEnricher.java]: What's the purpose of this?

File: test-framework/arquillian/src/main/java/io/quarkus/arquillian/MethodParameterInjectionPoint.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        return null;
+    }
+
+    public Member getMember() {
+        return method;
+    }
+
+    public Set<Annotation> getQualifiers() {
+        Set<Annotation> qualifiers = new HashSet<>();
+        // TODO - how to properly determine which annotations are qualifiers? beanManager.isQualifier() can't be used
```
Comment (mkouba) [inline test-framework/arquillian/src/main/java/io/quarkus/arquillian/MethodParameterInjectionPoint.java]: Filtering annotations annotated with `@Qualifier` should work in 90% use cases...

File: test-framework/arquillian/src/main/java/io/quarkus/arquillian/InjectionEnricher.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+                        .getMethod(method.getName(), convertToTCCL(method.getParameterTypes()));
+            } catch (Throwable t) {
+                throw new RuntimeException(t);
+            }
+            Class<?>[] parameterTypes = method.getParameterTypes();
+            for (int i = 0; i < parameterTypes.length; i++) {
+                try {
+                    values[i] = getInstanceByType(beanManager, i, method);
+                } catch (Exception e) {
+                    log.warn("CDIEnricher tried to lookup method parameter of type "
```
Comment (mkouba) [inline test-framework/arquillian/src/main/java/io/quarkus/arquillian/InjectionEnricher.java]: `CDIEnricher`? :-)

**Review by mkouba — CHANGES_REQUESTED** [review]
> I don't feel very comfortable with the convertToTCCL thingy but for some reason it was necessary (Arquillian is passing the method metadata to TestEnricher and TestMethodExecutor loaded by system class loader). Any better solution to go with?

It is needed. We use a different test instance due to quarkus-specific class loading issues.

File: test-framework/arquillian/src/main/java/io/quarkus/arquillian/InjectionEnricher.java (hunk 2 above)
Comment (jmartisk) [reply test-framework/arquillian/src/main/java/io/quarkus/arquillian/InjectionEnricher.java]: It's needed for getting injectable references, I basically took this from https://github.com/arquillian/arquillian-core/blob/1.4.1.Final/testenrichers/cdi/src/main/java/org/jboss/arquillian/testenricher/cdi/CDIInjectionEnricher.java#L103

Comment (manovotn) [reply test-framework/arquillian/src/main/java/io/quarkus/arquillian/InjectionEnricher.java]: I think Arquillian needs it for this - https://github.com/arquillian/arquillian-core/blob/1.4.1.Final/testenrichers/cdi/src/main/java/org/jboss/arquillian/testenricher/cdi/CreationalContextDestroyer.java#L36
Which looks like automagical CC cleanup :)

File: test-framework/arquillian/src/main/java/io/quarkus/arquillian/MethodParameterInjectionPoint.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+                qualifiers.add(potentialQualifier);
+            }
+        }
+        if (qualifiers.size() == 0) {
+            qualifiers.add(new DefaultLiteral());
+        }
+        return qualifiers;
+    }
+
+    private static class DefaultLiteral extends AnnotationLiteral<Default> implements Default {
```
Comment (manovotn) [inline test-framework/arquillian/src/main/java/io/quarkus/arquillian/MethodParameterInjectionPoint.java]: Shouldn't be needed from CDI 2.0
It is now built in - https://github.com/eclipse-ee4j/cdi/blob/master/api/src/main/java/javax/enterprise/inject/Default.java#L104

File: test-framework/arquillian/src/main/java/io/quarkus/arquillian/InjectionEnricher.java (hunk 2 above)
Comment (jmartisk) [reply test-framework/arquillian/src/main/java/io/quarkus/arquillian/InjectionEnricher.java]: Does that mean we should have a similar class in Quarkus too? We aren't importing that one

File: test-framework/arquillian/src/main/java/io/quarkus/arquillian/MethodParameterInjectionPoint.java (hunk 5 above)
Comment (jmartisk) [reply test-framework/arquillian/src/main/java/io/quarkus/arquillian/MethodParameterInjectionPoint.java]: If I remove that I am getting ambiguous dependencies on injecting the application `MetricRegistry` from https://github.com/smallrye/smallrye-metrics/blob/2.1.4/implementation/src/main/java/io/smallrye/metrics/MetricRegistries.java#L40

Comment (mkouba) [reply test-framework/arquillian/src/main/java/io/quarkus/arquillian/MethodParameterInjectionPoint.java]: You should use `Default.Literal.INSTANCE` instead...

Comment (jmartisk) [reply test-framework/arquillian/src/main/java/io/quarkus/arquillian/MethodParameterInjectionPoint.java]: Ok will use that one

Comment (jmartisk) [reply test-framework/arquillian/src/main/java/io/quarkus/arquillian/MethodParameterInjectionPoint.java]: Ah, ok

File: test-framework/arquillian/src/main/java/io/quarkus/arquillian/InjectionEnricher.java (hunk 2 above)
Comment (manovotn) [reply test-framework/arquillian/src/main/java/io/quarkus/arquillian/InjectionEnricher.java]: Aren't we? In that case why exactly did you need `InstanceProducer`? What consumers were there?
And yes, we should have a place where we invoke `CreationalContext.release()` to properly destroy all dependent instances linked to this `CreationalContext`.

Comment (jmartisk) [reply test-framework/arquillian/src/main/java/io/quarkus/arquillian/InjectionEnricher.java]: I guess you're right, in that case it wasn't quite necessary :) I added the `CreationalContextDestroyer` now

File: test-framework/arquillian/src/main/java/io/quarkus/arquillian/CreationalContextDestroyer.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+import javax.enterprise.context.spi.CreationalContext;
+
+import org.jboss.arquillian.core.api.Instance;
+import org.jboss.arquillian.core.api.annotation.Inject;
+import org.jboss.arquillian.core.api.annotation.Observes;
+import org.jboss.arquillian.core.spi.EventContext;
+import org.jboss.arquillian.test.spi.event.suite.After;
+
+public class CreationalContextDestroyer {
```
Comment (manovotn) [inline test-framework/arquillian/src/main/java/io/quarkus/arquillian/CreationalContextDestroyer.java]: Quarkus already has one Arq. observer registered - `QuarkusBeforeAfterLifecycle`.
We could just add the logic there.

Comment (manovotn) [reply test-framework/arquillian/src/main/java/io/quarkus/arquillian/CreationalContextDestroyer.java]: But we might want to rename that class, now that I think of it.
So it's just a matter of preference I think :)

Comment (jmartisk) [reply test-framework/arquillian/src/main/java/io/quarkus/arquillian/CreationalContextDestroyer.java]: I don't know, it feels that they are doing quite different things, so the split into two classes makes sense to me :)

**Review by manovotn — APPROVED** [review]

**Review by mkouba — APPROVED** [review]

**Pull request merged**